torch/_dynamo/guards.py

```diff
@@ -1247,71 +1241,105 @@ def replace(expr: str) -> Tuple[List[str], str]:
     return guard_body.getvalue(), make_guard_fn.getvalue()
 
 
-stashed_first_fail_reason = None
-
-
-def guard_fail_hook(
+def get_guard_fail_reason(
     guard_fn: GuardFn,
     code: types.CodeType,
     f_locals: Dict[str, object],
-    index: int,
-    last: bool,
-) -> None:
+) -> str:
     """
-    called whenever a guard fails.
+    Return the reason why `guard_fn` failed.
+    Updates `guard_failures` with the generated reason.
+    If `config.report_all_guard_failures` is enabled,
+    all failures in `guard_fn` are reported - otherwise,
+    only the first one is reported.
     """
-    first = index == 0
-    global stashed_first_fail_reason
-    # Don't waste time computing the fail reason for guards we aren't going to report out.
-    if not guard_fn.guard_fail_fn and not (first or last):
-        return
     scope = {"L": f_locals, "G": guard_fn.global_scope["G"]}
     scope.update(guard_fn.closure_vars)
     scope["___check_tensors"] = scope["___check_tensors_verbose"]
-    reason = ""
+    reasons: List[str] = []
     for part in guard_fn.code_parts:
         global_scope = dict(guard_fn.global_scope)
         global_scope["__compile_source__"] = part
         with report_compile_source_on_error():
-            fail_reason = eval(part, global_scope, scope)
+            fail_reason: Union[bool, str] = False
+            try:
+                fail_reason = eval(part, global_scope, scope)
+            except Exception:
+                # evaluating after the first failure can cause
+                # errors, e.g. if the check len(L['args']) > 0 fails,
+                # a subsequent check on L['args'][0] will error.
+                pass
         # Only ___check_tensors knows how to return a fancy fail reason;
         # for everything else we just report the code that failed
 
         if isinstance(fail_reason, bool) and not fail_reason:
             fail_reason = part
         if isinstance(fail_reason, str):
-            reason += fail_reason
-            if config.report_all_guard_failures:
-                reason += "\n"
-            else:
+            reasons.append(fail_reason)
+            if not config.report_all_guard_failures:
                 break
```

Comment:
Hmm, we may want to be a little careful about how exactly we do the reporting here.

There are a few levels of verbosity, it's not exactly only `report_all_guard_failures`. Here are the knobs:

* Do we report guard failures for all cache entries, or do we report for only one cache entry (and if only one, which one; want some sort of notion of best fit)
* For a single cache entry, do we want to report all the parts that failed, or only the first part that failed

By default, I think there are two desirable qualities for how we do reporting:

* What we report should correspond to what we actually happened in reality (so, e.g., by default we should NOT report all parts, because we didn't actually evaluate all of them at runtime)
* The report should give the information the user is most likely to want (so, e.g., in INFO, we should NOT report the failures for all the cache entries, because if you have a ton of cache entries like 64, that is going to be too much information for a typical user to want to know what to deal with).

Does that make sense? I'm not hard set on it but this is what I'm thinking about. I do think the extremely verbose variant can potentially be useful, and I think I would like to see it when debug is on, but for something lightweight like TORCH_LOGS=recompiles I'm guessing maybe it would be better to have something more condensed. IDK, have to dogfood it.